data_model/derive/src/lib.rs

```diff
@@ -14,216 +96,578 @@ use quote::{format_ident, quote};
-            .expect("Events should follow the naming format")
-    );
-
-    let event_variants = if let syn::Data::Enum(syn::DataEnum { variants, .. }) = data {
-        variants.iter().collect::<Vec<_>>()
-    } else {
-        panic!("Only `...Event` enums are supported")
-    };
-
-    let mut filter_variants_idents_id_fields = Vec::<syn::Ident>::new();
-    let mut filter_variants_idents_event_fields = Vec::<syn::Ident>::new();
-
-    let import_path = quote! { crate::prelude };
-
-    let filter_variants_event_fields = event_variants
-        .iter()
-        .filter_map(|variant| {
-            let event_filter_variant_ident = format_ident!("By{}", variant.ident);
-            if let syn::Fields::Unnamed(ref unnamed) = variant.fields {
-                let variant_type = &unnamed
-                    .unnamed
-                    .first()
-                    .expect("Should have at least one type")
-                    .ty;
-                process_event_variant(
-                    variant_type,
-                    &event_filter_variant_ident,
-                    &mut filter_variants_idents_id_fields,
-                    &mut filter_variants_idents_event_fields,
-                )
-                .map(|inner_filter_ident| {
-                    Some(quote! {
-                        #event_filter_variant_ident (#import_path::FilterOpt<#inner_filter_ident>)
-                    })
+mod id {
+    use super::*;
+
+    pub(super) struct IdInput {
+        _attrs: Vec<Attribute>,
+        _vis: Visibility,
+        _enum_token: Option<Token![enum]>,
+        _struct_token: Option<Token![struct]>,
+        ident: Ident,
+        _generics: Generics,
+        _brace_token: Brace,
+        _enum_variants: Option<Punctuated<Variant, Token![,]>>,
+        struct_fields: Option<Punctuated<Field, Token![,]>>,
+        id_type: TypePath,
+    }
+
+    impl Parse for IdInput {
+        fn parse(input: ParseStream) -> syn::Result<Self> {
+            let attrs = input.call(Attribute::parse_outer)?;
+            let id_type = parse_id_attribute(&attrs);
+            let vis = input.parse::<Visibility>()?;
+            if input.peek(Token![struct]) {
+                // Most usual case, most of the `domain_model` identifiables
+                let struct_token = Some(input.parse::<Token![struct]>()?);
+                let content;
+                Ok(IdInput {
+                    _attrs: attrs,
+                    _vis: vis,
+                    _enum_token: None,
+                    _struct_token: struct_token,
+                    ident: input.parse()?,
+                    _generics: input.parse()?,
+                    _brace_token: syn::braced!(content in input),
+                    struct_fields: Some(content.parse_terminated(Field::parse_named)?),
+                    _enum_variants: None,
+                    id_type,
                 })
             } else {
-                None
+                // Only used in [`IdBox`] case
```

Comment:
I get the feeling that implementing `Identifiable` on enums doesn't make sense? if you agree, I would rather make this implementation manual and remove the else branch